dbms/src/Operators/LocalAggregateRestorer.cpp

```diff
@@ -0,0 +1,152 @@
+// Copyright 2022 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Interpreters/Aggregator.h>
+#include <Operators/LocalAggregateRestorer.h>
+
+namespace DB
+{
+LocalAggregateRestorer::LocalAggregateRestorer(
+    const BlockInputStreams & bucket_streams,
+    Aggregator & aggregator_,
+    std::function<bool()> is_cancelled_,
+    const String & req_id)
+    : aggregator(aggregator_)
+    , is_cancelled(std::move(is_cancelled_))
+    , log(Logger::get(req_id))
+{
+    for (const auto & bucket_stream : bucket_streams)
+        bucket_inputs.emplace_back(bucket_stream);
+    if (bucket_inputs.empty())
+        finish();
+}
+
+void LocalAggregateRestorer::finish()
+{
+    assert(!finished);
+    finished = true;
+    bucket_inputs.clear();
+    LOG_INFO(log, "local agg restore finished");
+}
+
+bool LocalAggregateRestorer::loadFromInputs()
+{
+    assert(!bucket_inputs.empty());
+    for (auto it = bucket_inputs.begin(); it != bucket_inputs.end() && unlikely(!is_cancelled());)
+        it = it->load() ? std::next(it) : bucket_inputs.erase(it);
+    if unlikely (is_cancelled() || bucket_inputs.empty())
+    {
+        finish();
+        return false;
+    }
+    return true;
+}
+
+void LocalAggregateRestorer::storeFromInputToBucketData()
+{
+    assert(!bucket_inputs.empty());
+
+    // get min bucket num.
+    Int32 min_bucket_num = NUM_BUCKETS;
+    for (auto & bucket_input : bucket_inputs)
+        min_bucket_num = std::min(bucket_input.bucketNum(), min_bucket_num);
+    if unlikely (min_bucket_num >= NUM_BUCKETS)
+    {
+        finish();
+        return;
+    }
+
+    // store bucket data of min bucket num.
+    for (auto & bucket_input : bucket_inputs)
+    {
+        if (min_bucket_num == bucket_input.bucketNum())
+            bucket_data.push_back(bucket_input.moveOutput());
+    }
+    assert(!bucket_data.empty());
+}
+
+void LocalAggregateRestorer::loadBucketData()
```

Comment:
In executeIO, we will call read for all buckets. Is it expected?

Reply:
Each file written by the aggregator has 256 buckets and is ordered by bucket number.
Each file is read by a blockinputstream.
The logic here is to read only one bucket of each stream, and put buckets with the same bucket num together.
```
{bucket0, bucket1, ... bucket256} file1/input1
{bucket0, bucket1, ... bucket256} file2/input2
{bucket0, bucket1, ... bucket256} file3/input3
{bucket0, bucket1, ... bucket256} file4/input4
```
Did not read all the buckets in the file at once, so it is as expected

Reply:
However, it may take too long to execute here, because the disk will be read `file_num` times continuously, which can be optimized later.